src/libostree/ostree-repo-pull.c

```diff
@@ -3566,6 +3570,680 @@ ostree_repo_pull_with_options (OstreeRepo             *self,
+ * downloaded successfully after all remotes have been tried, %G_IO_ERROR_FAILED
+ * will be returned. The results of any successful downloads will remain cached
+ * in the local repository.
+ *
+ * If @cancellable is cancelled, %G_IO_ERROR_CANCELLED will be returned
+ * immediately. The results of any successfully completed downloads at that
+ * point will remain cached in the local repository.
+ *
+ * Returns: %TRUE on success, %FALSE otherwise
+ * Since: 2017.6
+ */
+gboolean
+ostree_repo_pull_from_remotes (OstreeRepo                            *self,
+                               const OstreeRepoFinderResult * const  *results,
+                               GVariant                              *options,
+                               OstreeAsyncProgress                   *progress,
+                               GCancellable                          *cancellable,
+                               GError                               **error)
+{
+  g_autoptr(GHashTable) refs_pulled = NULL;  /* (element-type utf8 gboolean) */
+  gsize i, j;
+  g_autoptr(GString) refs_unpulled_string = NULL;
+  GHashTableIter iter;
+  const gchar *ref;
+  gpointer is_pulled_pointer;
+
+  /* Keep track of the set of refs we’ve pulled already. Value is %TRUE if the
+   * ref has been pulled; %FALSE if it has not. */
+  refs_pulled = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
+
+  /* FIXME: Rework this code to pull in parallel where possible. At the moment
+   * we expect the (i == 0) iteration will do all the work (all the refs) and
+   * subsequent iterations are only there in case of error. */
+  for (i = 0; results[i] != NULL; i++)
+    {
+      const OstreeRepoFinderResult *result = results[i];
+      g_autoptr(GError) local_error = NULL;
+      g_autoptr(GVariant) local_options = NULL;
+      g_autoptr(GPtrArray) refs_to_pull = NULL;
+
+      refs_to_pull = g_ptr_array_new_with_free_func (NULL);
+
+      for (j = 0; result->refs[j] != NULL; j++)
+        if (!GPOINTER_TO_INT (g_hash_table_lookup (refs_pulled, result->refs[j])))
+          g_ptr_array_add (refs_to_pull, result->refs[j]);
+
+      if (refs_to_pull->len == 0)
+        {
+          g_debug ("Ignoring remote ‘%s’ as it has no relevent refs.", result->uri);
+          continue;
+        }
+
+      /* NULL terminator. */
+      g_ptr_array_add (refs_to_pull, NULL);
+
+      /* TODO: any other relevant options here? GPG verification? */
+      local_options = g_variant_new_parsed ("{"
+                                              "'override-url': <%s>,"
+                                              "'refs': <%as>,"
+                                            "}",
+                                            result->uri,
+                                            refs_to_pull->pdata);
+
+      /* TODO: Progress will yo-yo here. */
+      ostree_repo_pull_with_options (self, result->uri, local_options, progress,
```

Comment:
Yeah...would be *really* messy to try to drive multi-remote logic down farther unfortunately.